include/envoy/thread/thread.h

```diff
@@ -51,6 +51,27 @@ class ThreadFactory {
   virtual ThreadIdPtr currentThreadId() PURE;
 };
 
+#ifndef NDEBUG
```

Comment:
Thoughts on removing the NDEBUG guard? This is one case where it doesn't bother me to have this singleton. @jmarantz thoughts? Perhaps we should use the ref counting singleton thing you did for the tests?

Reply:
Not sure.

The benefit of using ./test/test_common/global.h is that you could actually get a fresh ThreadFactory on every test method, as opposed to re-using the instance from a previous test instance.

Of course since ThreadFactory implementations don't currently have state this is kind of moot. In the past I've worked with a thread system that maintained some state -- stats mostly and some maps for sanity checking, but we may never need that in Envoy.

The drawback of using ./test/test_common/global.h is that we'd be pulling global.h into source/... thus encouraging people to use it when we want to discourage singletons. I had created that helper class to solve a problem that I think was specific to tests and the 0-arg constructor mocking patterns.

I think on balance I'd rather leave global.h where it is, have the code here just reference a static pointer as Andres wrote it. I agree we might as well remove NDEBUG and just use it though, as @mattklein123 suggests.

Reply:
> Thoughts on removing the NDEBUG guard?

I'm fine with this. I figured we would want to discourage static access to the ThreadFactory given the current approach to singletons in general, but I can certainly remove the guard if we want to allow this use case as an exception to the rule.

> The drawback of using ./test/test_common/global.h is that we'd be pulling global.h into source/... thus encouraging people to use it when we want to discourage singletons.

I intentionally avoided a more generalized singleton solution for this reason. My goal with the current approach is to keep the implementation simple and focused on the ThreadFactory to discourage reuse of this pattern (for anything other than another special case).

Reply:
SGTM